### PR TITLE
Personality architecture refactor 

### LINKED ISSUE #28: "Personality" architecture overhaul discussion
Discussion from [here](https://github.com/cannawen/metric_units_reddit_bot/commit/37ee286dd390d32389ffc5916b7c441db23c8a94)

screen shot 2017-10-02 at 7 46 22 pm

### PR DESCRIPTION
**Pull request by chazzlabs**
## PR Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/cannawen/metric_units_reddit_bot/blob/master/CONTRIBUTING.md)
- [x] New functionality (or bug fix) is appropriately tested
- [x] PR includes `closes #{issue number}` if appropriate

## Brief description of changes:

This PR includes the changes made for the personality architecture overhaul.

All unit tests are passing; since the personality module now randomly chooses a personality from all personalities that match the message, I modified the existing personality unit tests to check the actual replies against the entire list of possible replies for a given message.

Notable changes included in this PR:
* `personality.js` was moved to `src/personality/index.js`
* All existing personalities defined in `personality.js` were removed and now exist in their own YAML configuration files in `src/personality/[robot, human]`
* Individual personality configurations now no longer include a `postprocess` function; there is a single `postprocess` function that is run on all responses, and the map which defines the supported values is returned in the `createPostprocessMap` function.
* `regex` properties in personality configurations can no longer be functions; instead they can be either a regex string or an array of regex strings.  In the case that a personality configuration's `regex` is an array of strings, a message is considered a match if it matches all of the strings in the array.

I apologize for the size of the changes in this PR, but it needs to be reviewed before being merged (plus I want that sweet t-shirt 😄 ).

Something to note: I originally had some detailed comments explaining the usage of the personality config files (the files that contained ordered lists for robot and human personalities).  Since those files no longer exist, where should we put any appropriate documentation for personalities?  Basically the documentation should give an overview of the properties contained in the YAML files for each personality (description, regex, responses, etc.).  I can add the documentation and commit again before the PR is merged if we decide the it should be included.

## Relevant link(s):

Resolves #28

### CODE REVIEW

File: src/personality/index.js
Comment (cannawen) [inline src/personality/index.js:20]: I'm wondering if we want to make all personalities case-insensitive? That way we don't need to include the `/i` in the .yaml files, and just have `"good(?: fellow)? human"` instead of `"/good(?: fellow)? human/i"` ... but it may reduce clarity/explicitness of the regex in the .yaml files. Hmm. Thoughts?

File: src/personality/index.js
Comment (cannawen) [inline src/personality/index.js:107]: Ohhh, this forEach looks much nicer than `Object.keys(dictionary).forEach`, good choice!

File: src/personality/index.js
Comment (cannawen) [inline src/personality/index.js]: I think this one should be `match.toUpperCase()` for "I'M A{{ADJECTIVE}} BOT"

File: src/personality/robot/sentient-bot.yaml
Comment (cannawen) [inline src/personality/robot/sentient-bot.yaml]: There should be a space before "elon" (or else it would match like... melon, lol)

File: test/personality/personality-test.js
Comment (cannawen) [inline test/personality/personality-test.js]: Now that we've pulled our personalities into data files, I am not sure if we should be testing them as thoroughly (since testing data files seems like overkill to me?). Do you think it would make sense to do just one thorough pass-through with a more complicated personality (like "who's a good bot")? Or perhaps it makes sense to fake out our "personality" file(s)?

And I was also wondering if we could set random to return 0 or 0.999, to test that randomness is actually working as expected, instead of using the true Math.random

This all sounds like a lot of work though, maybe we can make a different story to handle all this later?

**Review by cannawen — CHANGES_REQUESTED** [review]
Everything looks sooooo much better than before!! Great job with the refactor :D Just some small comments.

Also: I would like to wait until #90 and #97 are merged before putting this in, there will 100% be merge conflicts incoming (sorry)

**Comment by cannawen** [comment]
I do think there should be some documentation, perhaps in the [Contribute Code](https://github.com/cannawen/metric_units_reddit_bot/blob/master/CONTRIBUTING.md#contribute-code) section? Wherever it makes sense to you :)

File: src/personality/index.js
Comment (chazzlabs) [reply src/personality/index.js:107]: 👍 [lodash](https://github.com/lodash/lodash) is filled with utilities like this.

File: src/personality/index.js
Comment (chazzlabs) [reply src/personality/index.js]: Wow... Nice catch.

By the way, the reason I'm using `Object.assign` here is because the resulting object, when `match` is undefined, looks like:
```
{ username: 'foobar' }
```

Had I done something like
```
return {
    username: username,
    adjective: match !== undefined ? adjective.toLowerCase() : undefined,
    ADJECTIVE: match !== undefined ? adjective.toUpperCase() : undefined
};
```

the result would look like:
```
{
    username: 'foobar',
    adjective: undefined,
    ADJECTIVE: undefined
}
```

I didn't want those two additional properties to exist at all even if their values would be `undefined`.

File: test/personality/personality-test.js
Comment (chazzlabs) [reply test/personality/personality-test.js]: Yeah, it probably makes more sense and is more maintainable to just have well-crafted test input for the personalities; we don't necessarily want to test our specific personalities but rather whether the `personality` module properly selects responses for given input.  I agree that a different story is probably the better spot for that work.

I can also change the `Math.random` calls in `personality-test.js` to better test the randomness of responses.

**Comment by cannawen** [comment]
Still waiting for #105 to merged... I poked them earlier so hopefully it won't be too much longer :P

Are there more changes to this PR that you haven't pushed up yet?

**Comment by chazzlabs** [comment]
@cannawen  Are we still also waiting on #90 before merging this?  You mentioned waiting on that issue and #97 in a previous comment.  

I do still need to address the changes you brought up, but they should be quick and easy to knock out.

**Comment by cannawen** [comment]
Sorry for the confusion, #90 is the issue, and #105 is the PR. 

It's merged in now, so yay!

**Comment by chazzlabs** [comment]
@cannawen Just wanted to give you a heads up that this will be complete in the next day. Sorry for the delay on this.

**Comment by cannawen** [comment]
No worries! :)

**Comment by chazzlabs** [comment]
@cannawen I was finally able to address the changes for this PR.  I apologize again for the delay.

I'm not sure what the conflicts are.  I thought I'd addressed them for `package.json` and `src/bot.js`.  `src/personality.js` no longer exists in my branch, so I'm not sure about resolving that conflict.

Taking a look at the files in my branch, it looks like there should be no issues, right?

https://github.com/chazzlabs/metric_units_reddit_bot/blob/personality-refactor/package.json#L12
https://github.com/chazzlabs/metric_units_reddit_bot/blob/personality-refactor/src/bot.js#L30

One thing I didn't change, however, is the case (in)sensitivity for personality matches.  I'm not sure if we want to handle that as a change for this PR or as a separate issue.

**Comment by cannawen** [comment]
I fixed the merge conflicts (since the branch was behind master) :) Thanks!

The bot is slightly shadowbanned right now, I asked the admins why and hopefully it'll get unbanned. But we probably won't do any work while it's banned

**Pull request merged**
